File: geolocation/geolocation.py
'''
            BYU AUVSI-SUAS Capstone
         Target Geolocation Algorithm
              Connor Olsen, 2019
'''

import numpy as np
import math as math

"""
Calculates the meters between two GPS coordinates
@type lat1: float
@param lat1: The latitude of the first GPS coordinate

@type lon1: float
@param lon1: The longitude of the first GPS coordinate

@type lat2: float
@param lat2: The latitude of the second GPS coordinate

@type lon2: float
@param lon2: The longitude of the first GPS coordinate

@rtype: array of floats(4)
@return: Distance north (meters), distande east (meters), total distance (meters), angle (0 deg = East)
"""
def GPStoMeters(lat1, lon1, lat2, lon2):
    d2r = 0.0174532925199433
    dlong = (lon2 - lon1) * d2r
    dlat = (lat2 - lat1) * d2r
    a = (math.sin(dlat/2.0))**2 + math.cos(lat1*d2r) * math.cos(lat2*d2r) * (math.sin(dlong/2.0))**2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
    d = 6367 * c; #Distance between points in meters
    d_meters = d * 1000
    dy = lat2 - lat1
    dx = math.cos(d2r * lat1) * (lon2 - lon1)
    angle = math.atan2(dy, dx)
    east_dis_meters = d_meters * math.cos(angle)
    north_dis_meters = d_meters * math.sin(angle)
    returnvalues = np.array([north_dis_meters, east_dis_meters, d_meters, angle])
    return returnvalues

"""
Calculates GPS coordniates given a starting coordinate and meters north and east
@type Lat: float
@param Lat: The latitude of the GPS coordinate

@type Lon: float
@param Lon: The longitude of the GPS coordinate

@type north_displacement: float
@param north_displacement: The distance north of the given coordinates (meters)

@type east_displacement: float
@param east_displacement: The distance east of the given coordinates (meters)

@rtype: array of floats(2)
@return: latitude of new target, longitude of new target
"""
def MeterstoGPS(Lat, Lon, north_displacement, east_displacement):
    # Earth’s radius, sphere
    R = 6378137
    # Coordinate offsets in radians
    dLat = north_displacement/R
    dLon = east_displacement/(R*math.cos(math.pi*Lat/180))
    # OffsetPosition, decimal degrees
    latO = Lat + dLat * 180/math.pi
    lonO = Lon + dLon * 180/math.pi
    returnvals = [latO, lonO];
    return returnvals



'''
The data below will be pulled from the database:
Attitude
MAV Coordinates
Pixel Coordinates of Target
'''
# For now, we will use dummy data
phi_in = 0
theta_in = 60
psi_in = 90
alpha_az = 0 # Assuming the camera is angled with the top facing out the nose
alpha_el = -math.pi/2
lat_mav = 40.2465
lon_mav = -111.6483
lat_gnd = 40.2485
lon_gnd = -111.6458
height = 16

MaxX = 2000 # Max x pixels
MaxY = 2000 # Max y pixels
'''
The top left and bottom right coordinates of the cropped
photo are provided. This section finds the center of the
cropped image and translates it into
'''
TopLeftX = 0
TopLeftY = 0
BottomRightX = 0
BottomRightY = 0
CenterX = BottomRightX - TopLeftX
CenterY = BottomRightY - TopLeftY
AdjustedCenterX = CenterX-(MaxX/2)
AdjustedCenterY = (-1)*(CenterY-(MaxY/2))

M = 4000
Ex = -15 #AdjustedCenterX
Ey = -1028 #-AdjustedCenterY

fov_ang = 0.872665 #field of View angle --> A6000 83* - 32* (in radians)
f = M/(2*math.tan(fov_ang/2))

l_cusp_c = 1/math.sqrt(Ex**2 + Ey**2 + f**2) * np.array([[Ex],[Ey],[f]])
'''
Convert Roll, Pitch and Yaw to radians
'''
phi = phi_in*math.pi/180
theta = theta_in*math.pi/180
psi = psi_in*math.pi/180

'''
k unit vector in the inertial frame
'''
k_i = np.array([[0],[0],[1]])

'''
Calculates distance between ground station and mav in meters to determine MAV's
relative location. Then creates the position vector [Pn Pe Pd]^T
'''
positionData = GPStoMeters(lat_gnd, lon_gnd, lat_mav, lon_mav)
P_i_mav = np.array([[positionData[0]], [positionData[1]], [-height]])

'''
Trigonometry calculated one time to decrease run time
'''
cphi = math.cos(phi)
sphi = math.sin(phi)
ctheta = math.cos(theta)
stheta = math.sin(theta)
cpsi = math.cos(psi)
spsi = math.sin(psi)
caz = math.cos(alpha_az)
saz= math.sin(alpha_az)
cel = math.cos(alpha_el) #Rreturns 6.123234e-17 instead of 0
sel = math.sin(alpha_el)

'''
Rotation from body to inertial frame
Found on page 15 of Small Unmanned Aircraft
'''
R_v2b = np.array([[ctheta*cpsi, ctheta*spsi, -stheta],\
[sphi*stheta*cpsi-cphi*spsi, sphi*stheta*spsi+cphi*cpsi, sphi*ctheta],\
[cphi*stheta*cpsi+sphi*spsi, cphi*stheta*spsi-sphi*cpsi, cphi*ctheta]])
R_b2v = np.transpose(R_v2b)
R_b2i = R_b2v

'''
R_g_to_b
Found on page 227 of Small Unmanned Aircraft
'''
R_b2g1 = np.array([[caz, saz, 0],[-saz, caz, 0],[0, 0, 1]])
R_g12g = np.array([[cel, 0, -sel],[0, 1, 0],[sel, 0, cel]])
R_b2g = np.matmul(R_g12g, R_b2g1)
R_g2b = np.transpose(R_b2g)

'''
R_c_to_g
% Found on page 227 of Small Unmanned Aircraft
'''
R_g2c = np.array([[0, 1, 0],[0, 0, 1],[1, 0, 0]])
R_c2g = np.transpose(R_g2c)

'''
% For simplicity, the three Rotation matrices are combined into one below
RbiRbgRcg = R_b_to_i * R_g_to_b * R_c_to_g;
'''
RbiRbgRcg = np.matmul(np.matmul(R_b2i, R_g2b), R_c2g)


l_cusp_i = np.matmul(RbiRbgRcg, l_cusp_c)
P_i_tar = P_i_mav + height*l_cusp_i/l_cusp_i[2]

print("Groundstation coordinates")
print(str(lat_gnd) + " " + str(lon_gnd))

print("MAV coordinates")
print(str(lat_mav) + " " + str(lon_mav))

TargetCoordinates = MeterstoGPS(lat_gnd, lon_gnd, P_i_tar[0], P_i_tar[1])
print("Target coordinates")
print(str(float(TargetCoordinates[0])) + " " + str(float(TargetCoordinates[1])))

File: geolocation/geolocation.m
clear
clc
digits(100)
Coordinates % Load in the Test GPS Points
%% Initial Parameters
% These initial parameters represent data that will be readily accessible
% on the MAV. Pixel Array, Field of View angle are specifications of the
% camera, and the normalized line of sight vector (l_cusp_c)  will be
% determined by either the autonomous detection program, or the manual
% interface clients.

%TEST DATA
MAV_HEIGHT = 16;
X_PIXEL = -15;
Y_PIXEL = 1028;
ROLL = 0;
PITCH = 0.523599;
YAW = pi/2;
MAV_Coordinates = C.skybridge;
GroundStation_Coordinates = C.TestGroundStation;

M = 4000; %square pixel array (width and height if square)
Ex = X_PIXEL; %x_pixel/M; %X pixels, normalized
Ey = (-1)*Y_PIXEL; %y_pixel/M; %Y pixels, normalized
fov_ang = 0.872665;%sym('fov_ang'); %field of View angle --> A6000 83* - 32* (in radians)
f = M/(2*tan(fov_ang/2)); %focal length  in units of pixels

l_cusp_c = 1/sqrt(Ex^2 + Ey^2 + f^2) * [Ex;Ey;f];

roll = ROLL; %sym('roll'); %Radians
pitch = PITCH;%-0.523599; %sym('pitch'); %RadiansP
yaw = YAW;%pi/2; %sym('yaw'); %Radians

alpha_az = 0;%sym('az'); %Azmuth Angle: Should equal yaw.
alpha_el = -pi/2 + pitch;%sym('el'); %Elevation Angle: "Pitch" of the gimbal

k_i = [0;0;1];

MavGPSData = [GroundStation_Coordinates(1), GroundStation_Coordinates(2), MAV_Coordinates(1), MAV_Coordinates(2)];
MavLocationData = GPStoMeters(MavGPSData(1), MavGPSData(2), MavGPSData(3), MavGPSData(4));

Pn = MavLocationData(1);
Pe = MavLocationData(2);
Pd = (-1)*MAV_HEIGHT;%sym('Pd');
h = -Pd;%;sym('h');
%% Rotation Matrices
% The location of the target is defined in the camera coordinate frame
% system, and must be transformed into the inertial frame to provide GPS
% coordinates. This transformation is done using three rotation matrices:
% Moving Camera frame to Gimbal coordinates; Gimbal coordinates to body
% frame; body frame to inertial frame

% R_b_to_i
% Found on page 15 of Small Unmanned Aircraft

R_v_to_b = [cos(pitch)*cos(yaw) cos(pitch)*sin(yaw) -sin(pitch); ...
    sin(roll)*sin(pitch)*cos(yaw)-cos(roll)*sin(yaw) ...
    sin(roll)*sin(pitch)*sin(yaw)+cos(roll)*cos(yaw) sin(roll)*cos(pitch); ...
    cos(roll)*sin(pitch)*cos(yaw)+sin(roll)*sin(yaw) ...
    cos(roll)*sin(pitch)*sin(yaw)-sin(roll)*cos(yaw) cos(roll)*cos(pitch)];

R_b_to_v = R_v_to_b'; %R_v_b is a translation of R_i_b. It may work, but I may need one more step.
R_b_to_i = R_b_to_v;

% ------------------------------------------------------------------------
% R_g_to_b
% Found on page 227 of Small Unmanned Aircraft
R_b_to_g1 = [cos(alpha_az) sin(alpha_az) 0;...
    -sin(alpha_az) cos(alpha_az) 0;...
    0 0 1];

R_g1_to_g = [cos(alpha_el) 0 -sin(alpha_el);...
    0 1 0;...
    sin(alpha_el) 0 cos(alpha_el)];

R_b_to_g = R_g1_to_g*R_b_to_g1;
R_g_to_b = R_b_to_g';

% ------------------------------------------------------------------------
% R_c_to_g
% Found on page 227 of Small Unmanned Aircraft
R_g_to_c = [0 1 0;...
    0 0 1;...
    1 0 0];
R_c_to_g = R_g_to_c';

% For simplicity, the three Rotation matrices are combined into one below
RbiRbgRcg = R_b_to_i * R_g_to_b * R_c_to_g;

%% Geolocation Algorithm
% The algorithm uses the line of sight vector in the camera frame and
% rotates it into the inertial frame. There, it is multiplied by a scaler
% (height) and divided by the projection of the transformed line of sight
% vector onto the unit vector k_i in the inertial frame
P_i_mav = [Pn;Pe;Pd];
P_i_obj = P_i_mav + h*(RbiRbgRcg*l_cusp_c)/(dot(k_i,(RbiRbgRcg*l_cusp_c)));

%% Publish
% The coordinates of the groundstation are used as a reference for
% determining the relative position of the MAV, which is in turn used to
% find the relative position of the target. The coordinates for all three
% points are shown below.
disp("Groundstation Coordinates ");
fprintf(MavGPSData(1) + " " + MavGPSData(2) + "\n\n");

disp("MAV Coordinates ");
fprintf(MavGPSData(3) + " " + MavGPSData(4) + "\n\n");

TargetCoordinates = MeterstoGPS(LatTestGroundStation, LonTestGroundStation, P_i_obj(1), P_i_obj(2));
disp("Target Coordinates ");
fprintf(TargetCoordinates(1)+", "+TargetCoordinates(2) + "\n");